File: 22-generate-parentheses/generate-parentheses.java
import java.util.*;


class Solution {
    private static List<String> res;
    private static List<String> stack;

    public List<String> generateParenthesis(int n) {
        res = new ArrayList<>();
        stack = new ArrayList<>();

        backtrack(n, 0, 0);

        return res;
    }

    private static void backtrack(int n, int openN, int closeN) {
        if (openN == closeN && closeN == n) {
            res.add(String.join("", stack));
            return;
        }

        if (openN < n) {
            stack.add("(");
            backtrack(n, openN + 1, closeN);
            stack.remove(stack.size() - 1);
        }

        if (closeN < openN) {
            stack.add(")");
            backtrack(n, openN, closeN + 1);
            stack.remove(stack.size() - 1);
        }
    }
}

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def dfs(s, openN, closeN):
            if openN == closeN == n:
                res.append(s)
                return

            if openN < n:
                dfs(s + '(', openN + 1, closeN)
            
            if closeN < openN:
                dfs(s + ')', openN, closeN + 1)
        
        dfs("", 0, 0)
        return res

        # time O(n * 2^2n) (if there are n pairs, we have 2^2n possible strings, which is an upper bound. the real time complexity is lower. Each string includes a copy operation which takes O(n).
        # space: O(2n) = O(n)